lib/compat.php

```diff
@@ -24,20 +24,119 @@
 	 * @return WP_Block_Type|false The registered block type on success, or false on failure.
 	 */
 	function register_block_type_from_metadata( $path, $args = array() ) {
-		$file = trailingslashit( $path ) . 'block.json';
-		if ( ! file_exists( $file ) ) {
+		$metadata_file = trailingslashit( $path ) . 'block.json';
+		if ( ! file_exists( $metadata_file ) ) {
 			return false;
 		}
 
-		$metadata = json_decode( file_get_contents( $file ), true );
-		if ( ! is_array( $metadata ) ) {
+		$metadata = json_decode( file_get_contents( $metadata_file ), true );
+		if ( ! is_array( $metadata ) || empty( $metadata['name'] ) ) {
 			return false;
 		}
 
+		$settings          = array();
+		$property_mappings = array(
+			'title'           => 'title',
+			'category'        => 'category',
+			'parent'          => 'parent',
+			'icon'            => 'icon',
+			'description'     => 'description',
+			'keywords'        => 'keywords',
+			'attributes'      => 'attributes',
+			'supports'        => 'supports',
+			'styles'          => 'styles',
+			// Alias recommended in documentation to distinguish from assets.
+			'styleVariations' => 'styles',
```

Comment:
```suggestion
			'styleVariations' => 'styles',
			'textDomain' => 'text_domain',
```

Reply:
Basically the same mapping is happening here - https://github.com/WordPress/gutenberg/pull/21065/files#diff-fbf4591b5f3adc3368a5b41fab07fc52R233-R248

These field mappings need to be in sync.

Reply:
:wave: Disclaimer before my review: I'm coming at this without prior experience in the registration mechanism, so I'm going to probably make some basic questions,  some other questions that were already answered elsewhere, and may miss big pieces of context. Bare with me!

- I'm a bit confused about the `styles` and `styleVariations` keys. I've inspected some blocks and I see that, currently, they use the `styles` key to declare style variations within their settings. If the idea is to move those declarations to block.json, can we perhaps keep `styles` for the block style variations and use `stylesheet` (or `stylesheet_path`, etc) to declare the path to the stylesheet? I can see how using `styles` in the settings but `styleVariations` in the block.json can create some confusion.

- Related to the above: can't a block provide a stylesheet and style variations? In the current proposal, both are conflated so the latest overrides the former.

- If we do the above, it seems that we may be able to generate the PHP keys automatically from the JS ones? In the surface, it seems like we're transforming camelCase to snake_case.

Reply:
> I'm a bit confused about the `styles` and `styleVariations` keys. I've inspected some blocks and I see that, currently, they use the `styles` key to declare style variations within their settings. If the idea is to move those declarations to block.json, can we perhaps keep `styles` for the block style variations and use `stylesheet` (or `stylesheet_path`, etc) to declare the path to the stylesheet? I can see how using `styles` in the settings but `styleVariations` in the block.json can create some confusion.

It's a chicken or the egg type of issues. We can't start using `styleVariations` because it's not yet supported. Regardless of how we move forward it is and will be confusing because of the similarity between `style` and `styles` as it exists today. `block.json` is a brand new thing and as you already noticed we can use this opportunity to better express the intent they share by using different names. In fact, we bridge both JavaScript and PHP so inevitably we run into this issue with coding styles related to how keys are named: snake_case vs camelCase. This is also why we will have to keep those mappings between all those formats. I like your proposal to use the "path" part with assets. Actually we can continue using both when someone wants to use the script or style handle. I'm not convinced that `stylesheet` is a good idea though because the function that registers stylesheets is called [`wp_register_style`](https://developer.wordpress.org/reference/functions/wp_register_style/) and everywhere in PHP it follows the same naming schema.

> Related to the above: can't a block provide a stylesheet and style variations? In the current proposal, both are conflated so the latest overrides the former.

I said it's confusing. There is a single `style` or `editor_style` for stylesheets. `styles` is for style variations. The answer is yes! You can provide all 3 of them :)

> If we do the above, it seems that we may be able to generate the PHP keys automatically from the JS ones? In the surface, it seems like we're transforming camelCase to snake_case.

Can you expand on that? I might need an example to understand the idea.

Reply:
@nosolosw – https://github.com/WordPress/gutenberg/pull/22519/commits/1c5e7c550827f1c69970435f62e069a912a63db1 proposes handling for both script/style handles and paths.

Reply:
Let's remove all the confusion caused by `styleVariations` and stick with the pre-existing `styles`: 9f70451.

Reply:
I believe the original suggestion of mapping other camel-cased properties should still apply, though? i.e. `textDomain` -> `text_domain`.

I'm currently working on a separate pull request to add [context](https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-context.md) properties to the RFC document and block type REST controller. I'm inclined to try to bring this into alignment of snake-case. The current `WP_Block` implementation still references this as camel-case, and would need to change, likely with deprecation:

https://github.com/WordPress/gutenberg/blob/8eb46eefb8ea0fe8877b026e34803e0c05aefa9e/lib/class-wp-block.php#L121-L122

Reply:
In a related effort, the mapping becomes less of an issue as we use now `textdomain` and `styles` in all the places.

Reply:
I'm planning to consolidate `editorScript` and `editorScriptPath` into one field, by using npm local package notation:
```json
{
    "editorScript": "scipt-handle",
    "script": "file://./filename.js"
}
```

We discussed that in the past when working on RFC but I completely forgot about it. Props to @aduth for the reminder :)